nano/node/request_aggregator.cpp

```diff
@@ -0,0 +1,225 @@
+			// clang-format on
+		}
+	}
+}
+
+void nano::request_aggregator::stop ()
+{
+	nano::unique_lock<std::mutex> lock (mutex);
+	stopped = true;
+
+	lock.unlock ();
+	condition.notify_all ();
+
+	if (thread.joinable ())
+	{
+		thread.join ();
+	}
+}
+
+std::size_t nano::request_aggregator::size ()
+{
+	nano::unique_lock<std::mutex> lock (mutex);
+	return requests.size ();
+}
+
+bool nano::request_aggregator::empty ()
+{
+	return size () == 0;
+}
+
+std::vector<nano::block_hash> nano::request_aggregator::aggregate (nano::transaction & transaction_a, channel_pool & pool_a) const
+{
+	std::vector<nano::block_hash> to_generate;
+	std::vector<std::shared_ptr<nano::vote>> cached_votes;
+	for (auto const & hash_root : pool_a.hashes_roots)
+	{
+		auto find_votes (votes_cache.find (hash_root.first));
+		if (!find_votes.empty ())
+		{
+			cached_votes.insert (cached_votes.end (), find_votes.begin (), find_votes.end ());
+		}
+		else if (!hash_root.first.is_zero () && store.block_exists (transaction_a, hash_root.first))
+		{
+			to_generate.push_back (hash_root.first);
+		}
+		else if (!hash_root.second.is_zero ())
+		{
+			nano::block_hash successor (0);
+			// Search for block root
+			successor = store.block_successor (transaction_a, hash_root.second);
+			// Search for account root
+			if (successor.is_zero ())
+			{
+				nano::account_info info;
+				auto error (store.account_get (transaction_a, hash_root.second, info));
+				if (!error)
+				{
+					successor = info.open_block;
+				}
+			}
+			if (!successor.is_zero ())
+			{
+				auto find_successor_votes (votes_cache.find (successor));
+				if (!find_successor_votes.empty ())
+				{
+					cached_votes.insert (cached_votes.end (), find_successor_votes.begin (), find_successor_votes.end ());
+				}
+				else
+				{
+					to_generate.push_back (successor);
+				}
+				auto successor_block (store.block_get (transaction_a, successor));
+				assert (successor_block != nullptr);
+				nano::publish publish (successor_block);
+				pool_a.channel->send (publish);
+			}
+			else
+			{
+				stats.inc (nano::stat::type::requests, nano::stat::detail::requests_ignored);
+			}
+		}
+	}
+	// Unique votes
+	std::sort (cached_votes.begin (), cached_votes.end ());
+	cached_votes.erase (std::unique (cached_votes.begin (), cached_votes.end ()), cached_votes.end ());
+	for (auto const & vote : cached_votes)
+	{
+		nano::confirm_ack confirm (vote);
+		pool_a.channel->send (confirm);
+	}
+	stats.add (nano::stat::type::requests, nano::stat::detail::requests_cached, stat::dir::in, cached_votes.size ());
+	return to_generate;
+}
+
+void nano::request_aggregator::generate (nano::transaction & transaction_a, std::vector<nano::block_hash> hashes_a, std::shared_ptr<nano::transport::channel> & channel_a) const
+{
+	size_t generated_l = 0;
+	auto i (hashes_a.begin ());
+	auto n (hashes_a.end ());
+	while (i != n)
+	{
+		std::vector<nano::block_hash> hashes_l;
+		for (; i != n && hashes_l.size () < 12; ++i)
```

Comment:
I think 12 should be some sort of constant here and in voting.cpp as well

Reply:
Done as `nano::network::confirm_ack_hashes_max` . Only place that cannot see this is `nano/secure/common.cpp` -> `nano::vote::vote` -> `assert (blocks_a.size () <= 12);`